File: 백준/Gold/17298. 오큰수/오큰수.cc
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

int N;

int main() {
    cin >> N;
    vector<int> v(N), place(N, -1);

    for (int i = 0; i < N; i++) {
        cin >> v[i];
    }
    
    stack<int> temp;

    for (int idx = 0; idx < N; idx++) {
        while(!temp.empty() && v[temp.top()] < v[idx]) {
            place[temp.top()] = v[idx];
            temp.pop();
        }
        temp.push(idx);
    }

    for (auto elem : place) {
        cout << elem << " ";
    }

    return 0;
}

File: 백준/Gold/17298. 오큰수/오큰수.py
import sys
input = sys.stdin.readline

n = int(input())
arr = list(map(int, input().split()))
stack = []
place = [-1] * n

for idx in range(n):
    while stack and arr[stack[-1]] < arr[idx]:
        place[stack.pop()] = arr[idx]
    stack.append(idx)

print(*place)